File: frontend/src/stores/help.js
import { defineStore } from 'pinia'
import { ref, computed, shallowRef, watch } from 'vue'
import helpIndex from '@/data/helpIndex.json'
import { buildHelpSearch } from '@/utils/helpSearch'
import { i18n } from '@/i18n'

// Help centre state (#2792). Everything here is client-side: the articles are
// Markdown files from docs/help, compiled into the bundle as lazy `?raw` chunks,
// and the index (nav, TOC, search corpus) is the JSON built from them. No API,
// no store hydration, works offline and in the desktop build.
//
// The glob is eager:false — each article is its own chunk, so opening the help
// centre downloads the one article being read, not the whole manual. The keys
// are paths relative to this file; `bodyPath` in the localized article is
// relative to docs/help, hence the prefix when looking one up.
const RAW = import.meta.glob('../../../docs/help/**/*.md', { query: '?raw', import: 'default' })
const RAW_PREFIX = '../../../docs/help/'

// Platform scoping (#2795): the index carries the manual for both clients, and
// an article can be written for one of them only. The web shows the desktop
// text — including for a reader on a phone browser, because the split is
// «desktop site vs. app», not screen width: an Android screenshot on a page the
// reader is looking at in a browser would simply be a lie.
const WEB_ARTICLES = helpIndex.articles.filter((a) => (a.platforms || ['web']).includes('web'))

// Frontmatter is metadata for the index, not content — strip it before render.
const FRONTMATTER_RE = /^---\r?\n[\s\S]*?\r?\n---\r?\n?/

const DEFAULT_LOCALE = 'ru'

// localize resolves an article to the reader's language (#2809). A translation
// lives in `locales[<lang>]` beside the Russian original, not in its place, so
// the Russian article is always here even when a language has no translation
// yet — in that case the reader gets the Russian text with `translated: false`,
// and the UI says so, rather than a blank page. `bodyPath` is the Markdown this
// language renders: the translated web body, or the Russian one as a fallback.
function localize(article, lang) {
  const loc = lang !== DEFAULT_LOCALE ? article.locales?.[lang] : null
  if (!loc) {
    return { ...article, bodyPath: article.path, translated: lang === DEFAULT_LOCALE }
  }
  return {
    ...article,
    title: loc.title,
    category: loc.category,
    keywords: loc.keywords,
    headings: loc.headings,
    text: loc.text,
    updated: loc.updated || article.updated,
    bodyPath: loc.path,
    translated: true,
  }
}

export const useHelpStore = defineStore('help', () => {
  // The whole store follows the interface language: title, category, headings
  // and search corpus are the translated ones, and the body is fetched from the
  // translated file. Recomputed, not read once at import, so switching the
  // language in settings re-renders the manual without a reload.
  const lang = computed(() => i18n.global.locale.value)
  const localized = computed(() => WEB_ARTICLES.map((a) => localize(a, lang.value)))
  const searcher = computed(() => buildHelpSearch(localized.value))

  // shallowRef: the cache holds long strings keyed by `<lang>:<slug>` and is
  // replaced wholesale, so deep reactivity would only cost proxy overhead. The
  // language is part of the key — after switching it, the same slug must not
  // serve the body cached for the previous one.
  const cache = shallowRef(new Map())
  const current = ref(null) // slug of the open article
  const body = ref('') // its markdown, frontmatter stripped
  const loading = ref(false)
  const error = ref('')
  const query = ref('')

  const articles = computed(() => localized.value)

  // Nav tree: categories in index order (the builder already sorted articles by
  // category, then order — one shape on every language, translated captions), so
  // grouping the localized list by its category label keeps that order.
  const categories = computed(() => {
    const out = []
    for (const a of localized.value) {
      let cat = out[out.length - 1]
      if (!cat || cat.name !== a.category) out.push((cat = { name: a.category, articles: [] }))
      cat.articles.push(a)
    }
    return out
  })

  const results = computed(() => (query.value.trim() ? searcher.value(query.value) : []))

  // One-off search for callers with a query box of their own — global search
  // (#2794) runs it while the help page's own `query` keeps whatever the reader
  // last typed there.
  function find(term, limit) {
    return String(term || '').trim() ? searcher.value(term, limit) : []
  }

  const meta = computed(() => localized.value.find((a) => a.slug === current.value) || null)
  const headings = computed(() => meta.value?.headings || [])

  // True when the open article has no translation for the current language and
  // is showing the Russian fallback — the reader is told, not left guessing.
  const untranslated = computed(() => (meta.value ? meta.value.translated === false : false))

  // Language the screenshots follow (#2816). It tracks the article, not the
  // interface: on the Russian fallback the shots stay Russian, so the pictures
  // never disagree with the prose next to them.
  const bodyLang = computed(() => (untranslated.value ? DEFAULT_LOCALE : lang.value))

  const defaultSlug = computed(() => localized.value[0]?.slug || '')

  function bySlug(slug) {
    return localized.value.find((a) => a.slug === slug) || null
  }

  // Neighbours for the «дальше/назад» footer links, in reading order.
  const neighbours = computed(() => {
    const list = localized.value
    const i = list.findIndex((a) => a.slug === current.value)
    if (i < 0) return { prev: null, next: null }
    return { prev: list[i - 1] || null, next: list[i + 1] || null }
  })

  // Two readers at once (#2794): the /help page and the contextual drawer that
  // opens over a board. Each owns its own slug/body/loading/error, so opening a
  // hint from the board does not rewrite the article the help page is showing —
  // they share only the markdown cache, which is what actually costs anything.
  const page = { current, body, loading, error }

  const drawerSlug = ref('')
  const drawerBody = ref('')
  const drawerLoading = ref(false)
  const drawerError = ref('')
  const drawerShown = ref(false)
  const drawer = {
    current: drawerSlug,
    body: drawerBody,
    loading: drawerLoading,
    error: drawerError,
  }

  async function load(slug, into) {
    const article = bySlug(slug)
    if (!article) {
      into.current.value = slug
      into.body.value = ''
      into.error.value = i18n.global.t('help.error.notFound')
      return
    }
    into.current.value = slug
    into.error.value = ''
    const key = `${lang.value}:${slug}`
    const cached = cache.value.get(key)
    if (cached !== undefined) {
      into.body.value = cached
      return
    }
    const loader = RAW[RAW_PREFIX + article.bodyPath]
    if (!loader) {
      // The index and the files went out of sync — `make help-index` was not
      // re-run, or the file was deleted. Say so instead of rendering a blank.
      into.error.value = i18n.global.t('help.error.notInBundle')
      into.body.value = ''
      return
    }
    into.loading.value = true
    try {
      const raw = await loader()
      const md = String(raw).replace(FRONTMATTER_RE, '')
      const next = new Map(cache.value)
      next.set(key, md)
      cache.value = next
      // A slower load that finished after the reader moved on must not paint
      // over the article they are looking at now.
      if (into.current.value === slug) into.body.value = md
    } catch {
      if (into.current.value === slug) {
        into.error.value = i18n.global.t('help.error.loadFailed')
        into.body.value = ''
      }
    } finally {
      into.loading.value = false
    }
  }

  // Switching the interface language re-fetches whatever is open in the new
  // language: the cache is keyed by language, so without this the body would
  // keep the text of the language the reader just left.
  watch(lang, () => {
    if (current.value) load(current.value, page)
    if (drawerShown.value && drawerSlug.value) load(drawerSlug.value, drawer)
  })

  function open(slug) {
    return load(slug, page)
  }

  // The help centre is a modal over whatever the reader is doing (#2792), not a
  // page: opening it must not take anyone off their board. Shown first, loaded
  // after — the article is a lazy chunk, and on a cold cache waiting for it
  // would look like the menu item did nothing.
  const centerShown = ref(false)

  function openCenter(slug) {
    centerShown.value = true
    // No slug means «open the help centre», not «open the first article»: a
    // reader coming back mid-manual keeps their place.
    return open(slug || current.value || defaultSlug.value)
  }

  function closeCenter() {
    centerShown.value = false
  }

  const drawerMeta = computed(
    () => localized.value.find((a) => a.slug === drawerSlug.value) || null,
  )
  const drawerUntranslated = computed(() =>
    drawerMeta.value ? drawerMeta.value.translated === false : false,
  )
  const drawerBodyLang = computed(() =>
    drawerUntranslated.value ? DEFAULT_LOCALE : lang.value,
  )

  // Contextual help: show the panel first, then load. The article is a lazy
  // chunk, so waiting for it before opening would look like the ? button did
  // nothing on a cold cache.
  function openDrawer(slug) {
    drawerShown.value = true
    return load(slug || defaultSlug.value, drawer)
  }

  function closeDrawer() {
    drawerShown.value = false
  }

  return {
    articles,
    categories,
    current,
    body,
    loading,
    error,
    query,
    results,
    meta,
    headings,
    untranslated,
    bodyLang,
    neighbours,
    defaultSlug,
    bySlug,
    find,
    open,
    centerShown,
    openCenter,
    closeCenter,
    drawerShown,
    drawerSlug,
    drawerBody,
    drawerLoading,
    drawerError,
    drawerMeta,
    drawerUntranslated,
    drawerBodyLang,
    openDrawer,
    closeDrawer,
  }
})

File: android/app/src/main/java/website/msdnna/tessera/data/model/Help.kt
package website.msdnna.tessera.data.model

import com.google.gson.annotations.SerializedName

/**
 * The help-centre index (#2795) — the same `helpIndex.json` the web bundles,
 * built from `docs/help` by `frontend/scripts/build-help-index.mjs` and copied
 * into the APK's assets at build time (see `app/build.gradle`, copyHelpContent).
 *
 * It carries everything except the article bodies: navigation, per-article
 * table of contents, and the search corpus ([HelpArticle.text] is the flattened
 * body). The Markdown itself is read from assets only when an article is opened.
 *
 * Lives in `data/model` on purpose: Gson fills these by reflection and the
 * package is the one `proguard-rules.pro` keeps. A model outside it survives
 * debug and comes back stripped in release — empty articles, no crash.
 */
data class HelpIndex(
    @SerializedName("articles") val articles: List<HelpArticle> = emptyList(),
)

data class HelpArticle(
    @SerializedName("slug") val slug: String = "",
    /** Path of the Markdown file relative to `docs/help` (assets: `help/<path>`). */
    @SerializedName("path") val path: String = "",
    @SerializedName("title") val title: String = "",
    @SerializedName("category") val category: String = "",
    @SerializedName("order") val order: Int = 0,
    @SerializedName("updated") val updated: String = "",
    @SerializedName("keywords") val keywords: List<String> = emptyList(),
    /** Clients the article is written for — `web`, `android`, or both. */
    @SerializedName("platforms") val platforms: List<String> = emptyList(),
    @SerializedName("headings") val headings: List<HelpHeading> = emptyList(),
    /** Flattened body, for search and result excerpts — not for rendering. */
    @SerializedName("text") val text: String = "",
    /** The mobile rewrite of this article, when one exists (`<slug>.android.md`). */
    @SerializedName("android") val android: HelpVariant? = null,
    /**
     * Translations of this article (#2809), keyed by language — `en`, and any
     * language added later. The Russian article always stays on the top level;
     * a translation lives here beside it, never in its place, so an older APK
     * whose index predates a language keeps showing the Russian help rather than
     * an empty screen.
     */
    @SerializedName("locales") val locales: Map<String, HelpLocale> = emptyMap(),
) {
    /**
     * Whether this article belongs in the app at all. An index without the field
     * (an older build's assets) counts as «both» rather than «neither»: showing
     * the desktop text beats an empty help section.
     */
    val onAndroid: Boolean
        get() = platforms.isEmpty() || platforms.contains(PLATFORM_ANDROID)

    /**
     * Collapses the article to what this client renders for [lang]: the
     * translated mobile rewrite when one exists, then the translated web body,
     * then the Russian mobile rewrite, then the Russian web body. Title and
     * category follow the same language. A language with no translation yet
     * falls back to Russian with [HelpContent.translated] false, so the screen
     * can say the article is not translated instead of hiding it.
     */
    fun content(lang: String): HelpContent {
        val loc = if (lang != DEFAULT_LANG) locales[lang] else null
        if (loc != null) {
            val v = loc.android
            return HelpContent(
                slug = slug,
                title = loc.title.ifBlank { title },
                category = loc.category.ifBlank { category },
                path = v?.path ?: loc.path,
                text = v?.text ?: loc.text,
                keywords = v?.keywords ?: loc.keywords,
                headings = v?.headings ?: loc.headings,
                updated = (v?.updated?.takeIf { it.isNotBlank() } ?: loc.updated).ifBlank { updated },
                translated = true,
                mobileRewrite = v != null,
            )
        }
        return HelpContent(
            slug = slug,
            title = title,
            category = category,
            path = android?.path ?: path,
            text = android?.text ?: text,
            keywords = android?.keywords ?: keywords,
            headings = android?.headings ?: headings,
            updated = android?.updated?.takeIf { it.isNotBlank() } ?: updated,
            translated = lang == DEFAULT_LANG,
            mobileRewrite = android != null,
        )
    }

    companion object {
        const val PLATFORM_ANDROID = "android"

        /** The language the manual is authored in; every other one is a translation. */
        const val DEFAULT_LANG = "ru"
    }
}

/**
 * The platform-specific half of an article (#2795): a different body, and with
 * it a different table of contents and search corpus. Title, category and place
 * in the navigation stay with the base article — the two platforms share one
 * manual structure and diverge only in the text.
 */
data class HelpVariant(
    /** Path of the variant's Markdown, relative to `docs/help`. */
    @SerializedName("path") val path: String = "",
    @SerializedName("updated") val updated: String = "",
    @SerializedName("keywords") val keywords: List<String> = emptyList(),
    @SerializedName("headings") val headings: List<HelpHeading> = emptyList(),
    @SerializedName("text") val text: String = "",
)

/**
 * A translation of an article (#2809). It carries the same shape as the base
 * article — its own title/category, web body and (when the mobile client has a
 * rewrite of it) an `android` half — so switching language swaps the whole
 * article, not just its prose.
 */
data class HelpLocale(
    @SerializedName("path") val path: String = "",
    @SerializedName("title") val title: String = "",
    @SerializedName("category") val category: String = "",
    @SerializedName("updated") val updated: String = "",
    @SerializedName("keywords") val keywords: List<String> = emptyList(),
    @SerializedName("headings") val headings: List<HelpHeading> = emptyList(),
    @SerializedName("text") val text: String = "",
    @SerializedName("android") val android: HelpVariant? = null,
)

/**
 * An article collapsed to a single language and platform (#2809) — the fields
 * the screen and the searcher actually use, with the locale/mobile fallbacks
 * already resolved. Not deserialized; produced by [HelpArticle.content].
 */
data class HelpContent(
    val slug: String,
    val title: String,
    val category: String,
    /** Markdown path relative to `docs/help`, unique per language and platform —
     *  so it doubles as the body cache key, and one language cannot serve the
     *  body cached for another. */
    val path: String,
    val text: String,
    val keywords: List<String>,
    val headings: List<HelpHeading>,
    val updated: String,
    /** False when a non-Russian reader is seeing the Russian original because no
     *  translation exists yet — the screen shows a note. */
    val translated: Boolean,
    /** True when [text]/[path] are the mobile rewrite; false when they are the
     *  desktop web body and the reader is warned it describes a mouse. */
    val mobileRewrite: Boolean,
)

data class HelpHeading(
    @SerializedName("id") val id: String = "",
    @SerializedName("text") val text: String = "",
    @SerializedName("level") val level: Int = 2,
)
